### ISSUE TITLE
Clojure LSP's Goto Definition/Show Documentation/et. issue in ClojureScript and namespace aliases (or macros??)

### ISSUE THREAD

**Issue by stevenproctor** [clj-kondo/clj-kondo#1372]
**version**

```
clojure-lsp 2021.08.24-14.41.56
clj-kondo 2021.08.07-SNAPSHOT
```
**macro usage**

Maybe???

I am seeing this on ClojureScript, and it might be related to macros being used there, since they are done as a different phase than Clojure.

I am not sure if it the macros themselves or specifically the namespace aliasing and rewriting might be causing things to be odd.

**platform**

I am running clj-kondo via clojure-lsp.

**editor**

NeoVim with its native LSP client, and clojure-lsp

**problem**

I am unable to view definitions, docstrings, or jump to definitions of some vars in ClojureScript.  These may be related to macros, but they may also be related to some namespace aliasing.

**repro**

An example project here with basic code and some screenshots:

https://github.com/stevenproctor/cljs-ns-aliasing-troubleshooting

In the `imgs` folder I have included screenshots of what I am seeing in my editor for different versions that are working (or not) as expected.

**config**

[ Is your bug related to `.clj-kondo/config.edn`? Paste your entire
configuration here. Also check if the bug still occurs without using a
configuration, as the bug may be a result of malformed configuration. ]

**expected behavior**

I would expect that I can see the documentation, arguments, or goto definition of the functions would work.  I also use Conjure for NeoVim, by Olical, and if attached to an NREPL, I am able to inspect the documentation via the REPL.

[ Optional: if you are interested in doing a PR yourself, please leave a note. ]
If after looking at this, it becomes some low hanging fruit, I would not mind taking a shot at the PR if that would help.

**Comment by borkdude** [comment]
As one data point, the `cljs-async/<!` goto definition does work for me, but that's just a normal function.

![cljs-async](https://user-images.githubusercontent.com/284934/133454816-bf324ec2-5c5d-455d-b875-5e655e764c0d.gif)

From a `.clj` file the navigation works correctly for me with the `go` macro:

```
(ns cljs-ns-aliasing-troubleshooting.src.cljs-ns-aliasing.core
  (:require [clojure.core.async :as async]))

(async/go)
```

But not when done from a `.cljs` file. Digging further.

**Comment by borkdude** [comment]
The namespace + var name is correct in the analysis:

```
   :name go,
   5   │   :filename "src/cljs_ns_aliasing/core.cljs",
   6   │   :alias async,
   7   │   :from cljs-ns-aliasing.core,
   8   │   :col 4,
   9   │   :name-col 4,
  10   │   :from-var hello-world,
  11   │   :arity 1,
  12   │   :row 8,
  13   │   :to clojure.core.async
  ```
  
  @ericdallo Could you have a look why the navigation isn't working?

**Comment by ericdallo** [comment]
Sure, I'll take a look with latest clojure-lsp

**Comment by ericdallo** [comment]
@borkdude it seems that clojure-lsp is ignoring the var-definition of that `go` macro for cljs because the `:filename` ends with .clj not cljs or cljc, not sure if that is suppose to work with cljs 🤔 
```clojure
{:end-row 30
 :name-end-col 13
 :name-end-row 12
 :name-row 12
 :ns cljs.core.async
 :name go
 :defined-by clojure.core/defmacro
 :filename "/home/greg/.m2/repository/org/clojure/core.async/1.3.618/core.async-1.3.618.jar:cljs/core/async.clj"
 :macro true
 :col 1
 :name-col 11
 :end-col 10
 :arglist-strs ["[& body]"]
 :varargs-min-arity 0
 :doc "..."
 :bucket :var-definitions
 :row 12}
```
and clojure-lsp don't consider clj files for cljs files: https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/queries.clj#L108

**Comment by borkdude** [comment]
@ericdallo It should in case of a macro.

**Comment by ericdallo** [comment]
@borkdude so you think clojure-lsp should match the file lang unless it's a macro always?

**Comment by borkdude** [comment]
@ericdallo There are a few rules around this.

From a .clj file, a function/macro can either come from a .clj or .cljc file
From a .cljs file, a function/macro can either come from a .cljs, .cljc or .clj file
From a .cljc file, similar rules as the above, but also depends on reader branch.

This logic is already implemented in clj-kondo [here](https://github.com/clj-kondo/clj-kondo/blob/cd86393d86bbe5e990e89c40e7f44d36cbd64b10/src/clj_kondo/impl/utils.clj#L238-L290).

Perhaps it's just a matter of adding something to the analysis data that can help you prevent re-implementing this logic. See impl/linters.clj where this resolve logic is called. I think after that, the call is inserted into the analysis, so perhaps we should add something there.

**Comment by ericdallo** [comment]
yeah, I think having `:available-langs` for those cases would help, something that tells that this analysis should work for those langs, WDYT?

**Comment by borkdude** [comment]
I've looked into it a bit but I think we can add some info to utils/resolve using

```
;; cljs func in another cljc file
(some-> (get-in idacs [:cljc :defs fn-ns :cljs fn-name])
  (assoc :call-lang :cljs :call-cljc true))
```

or so. Feel free to experiment, it's complicated so it would require me some time as well, perhaps I can look a bit in the weekend.

**Comment by borkdude** [comment]
@ericdallo See branch `add-call-ctx`, I added some info there:

borkdude@MBP2019 ~/Dropbox/dev/clojure/clj-kondo (add-call-ctx) $ clojure -M:clj-kondo/dev --lint - --lang cljs <<< "(require '[cljs.core.async :as a]) (a/go)"
:called-fn {:row 2962, :col 1, :macro true, :varargs-min-arity 0, :name require, :ns cljs.core, :top-ns cljs.core, :lang :cljs, :cljc true}

**Comment by borkdude** [comment]
Whoops, no the above example isn't complete, but hope that helps somewhat.

**Comment by borkdude** [comment]
So this is the complete output:

```
$ clojure -M:clj-kondo/dev --lint - --lang cljs <<< "(require '[cljs.core.async :as a]) (a/go)"
:called-fn {:ns cljs.core, :name require, :lang :cljs, :macro true, :col 1, :top-ns cljs.core, :cljc true, :varargs-min-arity 0, :row 2973}
:called-fn {:row 12, :col 1, :macro true, :varargs-min-arity 0, :name go, :ns cljs.core.async, :top-ns cljs.core.async, :lang :clj}
```
It shows that the call came from `:lang :clj` which must be a clojure file if `:cljc true` isn't there.
We can then add this to the analysis.

**Comment by borkdude** [comment]
We can also add the `:filename` perhaps:

```
                           ;; maybe a macro?
                           (when-let [ns (get-in idacs [:clj :defs fn-ns])]
                             (some-> (get-in idacs [:clj :defs fn-ns fn-name])
                                     (assoc :filename (:filename ns))))
```

**Comment by ericdallo** [comment]
@borkdude I'll take a look at that branch, what's the difference between `:ns` and `:top-ns`?
Also, it's not clear to me, clojure-lsp should check both `:lang` and `:cljc`?

**Comment by borkdude** [comment]
These are just proposals, not yet in the analysis data. `:top-ns` means the first ns declared in the file, clojure supports having multiple nss in one file.

**Comment by borkdude** [comment]
@ericdallo Let me just summarize it: 

What would you need in the analysis data. We could have:

- the `:filename` 
- some extra keys like `:cljc true`, or `:lang :clj`, or whatever (the exact format I must think about how well it fits with the already existing analysis data).

Alternative: you could just try in the logical order I described above (especially when `:macro true` you should look into .clj and .cljc).

**Comment by ericdallo** [comment]
Oh got it the `top-ns`, I see, so IMO we should tag with some key that would say where this analysis makes sense. for example if the analysis has a `:available-langs #{:cljc :cljs}`, that means that that analysis element can be used on analysis for clj + cljs files

**Comment by borkdude** [comment]
That's actually not necessary: clj-kondo will already resolve this for its own linting so it can just stick the file or extension as data in the analysis. So it's just a matter of what's most convenient for you, while also fitting it with the rest of the data.

**Comment by ericdallo** [comment]
ATM clojure-lsp uses this logica order you described (just not checking clj for cljs files), I think if clj-kondo always provides something that says which langs are available would be great, I mean, there are some analysis contains `:lang` but there are others that doesn't, that's why clojure-lsp also checks the `:filename` extension if no `:lang` here: https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/queries.clj#L43-L49

**Comment by borkdude** [comment]
Is there still something needed from this issue?
